server/sonar-server/src/main/java/org/sonar/server/qualityprofile/index/ActiveRuleIndexer.java

```diff
@@ -0,0 +1,155 @@
+    DbSession dbSession = dbClient.openSession(false);
+    long maxDate;
+    try {
+      ActiveRuleResultSetIterator rowIt = ActiveRuleResultSetIterator.create(dbClient, dbSession, lastUpdatedAt);
+      maxDate = doIndex(bulk, rowIt);
+      rowIt.close();
+      return maxDate;
+    } finally {
+      dbSession.close();
+    }
+  }
+
+  private long doIndex(BulkIndexer bulk, Iterator<ActiveRuleDoc> activeRules) {
+    bulk.start();
+    long maxDate = 0L;
+    while (activeRules.hasNext()) {
+      ActiveRuleDoc activeRule = activeRules.next();
+      bulk.add(newIndexRequest(activeRule));
+
+      // it's more efficient to sort programmatically than in SQL on some databases (MySQL for instance)
+      maxDate = Math.max(maxDate, activeRule.updatedAt());
+    }
+    bulk.stop();
+    return maxDate;
+  }
+
+  public void index(List<ActiveRuleChange> changes) {
+    deleteKeys(FluentIterable.from(changes)
+      .filter(MatchDeactivatedRule.INSTANCE)
+      .transform(ActiveRuleChangeToKey.INSTANCE)
+      .toList());
+    index();
+  }
+
+  public void deleteProfile(String qualityProfileKey) {
+    BulkIndexer bulk = new BulkIndexer(esClient, INDEX);
+    bulk.start();
+    SearchRequestBuilder search = esClient.prepareSearch(INDEX)
+      .setTypes(TYPE_ACTIVE_RULE)
+      .setQuery(QueryBuilders.filteredQuery(
+        QueryBuilders.matchAllQuery(),
+        FilterBuilders.boolFilter().must(FilterBuilders.termsFilter(RuleIndexDefinition.FIELD_ACTIVE_RULE_PROFILE_KEY, qualityProfileKey))
+        ));
+    bulk.addDeletion(search);
+    bulk.stop();
+  }
+
+  private void deleteKeys(List<ActiveRuleKey> keys) {
+    BulkIndexer bulk = new BulkIndexer(esClient, INDEX);
+    bulk.start();
+    SearchRequestBuilder search = esClient.prepareSearch(INDEX)
+      .setTypes(TYPE_ACTIVE_RULE)
+      .setQuery(QueryBuilders.filteredQuery(
+        QueryBuilders.matchAllQuery(),
+        FilterBuilders.boolFilter().must(FilterBuilders.termsFilter(RuleIndexDefinition.FIELD_ACTIVE_RULE_KEY, keys))
+        ));
+    bulk.addDeletion(search);
+    bulk.stop();
+  }
+
+  private BulkIndexer createBulkIndexer(boolean large) {
+    BulkIndexer bulk = new BulkIndexer(esClient, INDEX);
+    bulk.setLarge(large);
+    return bulk;
+  }
+
+  private IndexRequest newIndexRequest(ActiveRuleDoc doc) {
+    return new IndexRequest(INDEX, TYPE_ACTIVE_RULE, doc.key().toString())
+      .parent(doc.key().ruleKey().toString())
+      .routing(doc.key().ruleKey().repository())
+      .source(doc.getFields());
+  }
+
+  private enum MatchDeactivatedRule implements Predicate<ActiveRuleChange> {
+    INSTANCE;
+
+    @Override
+    public boolean apply(@Nonnull ActiveRuleChange input) {
+      return input.getType().equals(ActiveRuleChange.Type.DEACTIVATED);
+    }
+  }
+
+  private enum ActiveRuleChangeToKey implements Function<ActiveRuleChange, ActiveRuleKey> {
+    INSTANCE;
+
+    @Override
+    public ActiveRuleKey apply(@Nonnull ActiveRuleChange input) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The "input" parameter nullability is different in the superclass method, and that should not be changed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)